src/t_hash.c

```diff
@@ -2267,29 +2965,976 @@ void hpersistCommand(client *c) {
+                    /* If none of EX,EXAT,PX,PXAT,KEEPTTL is specified, TTL is
+                     * discarded. */
+                    listpackExPersist(o, field, fptr, vptr);
+                } else if (checkAlreadyExpired(expireAt)) {
+                    hashTypeDelete(o, field);
+                } else {
+                    if (*minPrevExp > expireAt)
+                        *minPrevExp = expireAt;
+
+                    listpackExUpdateExpiry(o, field, fptr, vptr, expireAt);
+                }
+            }
+            return 1;
+        }
+    } else if (o->encoding == OBJ_ENCODING_HT) {
+        hfield hf = NULL;
+        dictEntry *de = NULL;
+        dict *d  = o->ptr;
+        dictExpireMetadata *meta = (dictExpireMetadata *) dictMetadata(d);
+        sds prevVal = NULL;
+
+        /* First retrieve the field to check if it exists */
+        de = dictFind(d, field);
+        if (de) {
+            hf = dictGetKey(de);
+            prevExpire = hfieldGetExpireTime(hf);
+            prevVal = dictGetVal(de);
+        }
+
+        /* Check DCF (don't create fields) and DOF (don't override fields) arg. */
+        if ((!de && (flag & HFE_CMD_DCF)) || (de && (flag & HFE_CMD_DOF))) {
+            hsetfAddReply(c, flag, prevVal, prevVal, ret);
+            return 0;
+        }
+
+        /* Field value will be updated. */
+        ret = HSETF_FIELD;
+
+        /* Decide if we are going to set/discard TTL */
+        if (hsetfCheckTTLCondition(flag, prevExpire, expireAt))
+            ret = HSETF_FIELD_AND_TTL;
+
+        hsetfAddReply(c, flag, prevVal, value, ret);
+
+        if (!hf || !hfieldIsExpireAttached(hf)) {
+            hfieldFree(hf);
+
+            int withExpireMeta = (ret == HSETF_FIELD_AND_TTL) ? 1 : 0;
+            hf = hfieldNew(field, sdslen(field), withExpireMeta);
+
+            if (!de) {
+                dictUseStoredKeyApi(d, 1);
+                de = dictAddRaw(d, hf, NULL);
+                dictUseStoredKeyApi(d, 0);
+            }
+            dictSetKey(d, de, hf);
+        }
+
+        dictSetVal(d, de, sdsdup(value));
+        sdsfree(prevVal);
+
+        if (ret == HSETF_FIELD_AND_TTL) {
+            if (*minPrevExp > prevExpire)
+                *minPrevExp = prevExpire;
+
+            if (!(flag & HFE_CMD_EXPIRY_MASK)) {
+                /* If none of EX,EXAT,PX,PXAT,KEEPTTL is specified, TTL is
+                 * discarded. */
+                hfieldPersist(o, hf);
+            } else if (checkAlreadyExpired(expireAt)) {
+                /* if expiration time is in the past */
+                hashTypeDelete(o, field);
+            } else {
+                if (*minPrevExp > expireAt)
+                    *minPrevExp = expireAt;
+
+                if (prevExpire != EB_EXPIRE_TIME_INVALID)
+                    ebRemove(&meta->hfe, &hashFieldExpireBucketsType, hf);
+
+                ebAdd(&meta->hfe, &hashFieldExpireBucketsType, hf, expireAt);
+            }
+        }
+    }
+
+    return 1;
```

Comment:
we should make sure that hsetfSetFieldAndReply() takes with a reply, but not here.
if it doesn't enter, i think it can be changed to the following.
```suggestion
    redis_unreachable();
```

Reply:
Added assert to verify type. Also, changed where we return.  Hope it is clear.